File: CS/DiagramFromTree/Controllers/HomeController.cs
using DevExpress.Web.Mvc;
using DiagramFromTree.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace DiagramFromTree.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            return View(DepartmentProvider.GetDepartments());
        }

        public ActionResult LinearUpdate(MVCxDiagramNodeUpdateValues<Department, int> nodeUpdateValues)
        {
            foreach (var item in nodeUpdateValues.Update)
                DepartmentProvider.Update(item);
            foreach (var itemKey in nodeUpdateValues.DeleteKeys)
                DepartmentProvider.Delete(itemKey);
            foreach (var item in nodeUpdateValues.Insert)
            {
                var insertedItem = DepartmentProvider.Insert(item);
                nodeUpdateValues.MapInsertedItemKey(item, insertedItem.ID);
            }
            return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues);
        }
    }
}

File: VB/DiagramFromTreeVB/Controllers/HomeController.vb
Imports DevExpress.Web.Mvc

Public Class HomeController
    Inherits System.Web.Mvc.Controller

    Function Index() As ActionResult
        Return View(DepartmentProvider.GetDepartments())
    End Function

    Public Function LinearUpdate(ByVal nodeUpdateValues As MVCxDiagramNodeUpdateValues(Of Department, Integer)) As ActionResult
        For Each item In nodeUpdateValues.Update
            DepartmentProvider.Update(item)
        Next

        For Each itemKey In nodeUpdateValues.DeleteKeys
            DepartmentProvider.Delete(itemKey)
        Next

        For Each item In nodeUpdateValues.Insert
            Dim insertedItem = DepartmentProvider.Insert(item)
            nodeUpdateValues.MapInsertedItemKey(item, insertedItem.ID)
        Next

        Return DiagramExtension.GetBatchUpdateResult(nodeUpdateValues)
    End Function
End Class